pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/JsonExtractScalarTransformFunction.java

```diff
@@ -184,8 +191,7 @@ public long[] transformToLongValuesSV(ValueBlock valueBlock) {
       if (result instanceof Number) {
         _longValuesSV[i] = ((Number) result).longValue();
       } else {
-        // Handle scientific notation
-        _longValuesSV[i] = (long) Double.parseDouble(result.toString());
+          _longValuesSV[i] = Long.parseLong(result.toString());
```

Comment:
Just as a reminder, we are going to lose the ability to parse scientific notation, which is part of the [JSON standard](https://www.json.org/json-en.html).

Reply:
JSON standard does allow that for numbers but not for integers.
It's a question of whether we want to keep supporting conversion from json floating point ( scientific notation or a number with fractional part) string directly to long.

In terms of testing one can simulate it by adding the following to 
`JsonExtractScalarTransformFunctionTest.testJsonPathTransformFunctionDataProvider` :
```
testArguments.add(new Object[]{
          String.format("jsonExtractScalar(%s,'$.stringSV','LONG')", input), DataType.LONG, true
      });
```
I'd really avoid allocating object per row if result is going to be long.

Reply:
> JSON standard does allow that for numbers but not for integers.

That is not precise. In JSON there is only one type of numeric literal: _number_. A _number_ is defined as `number fraction exponent`. Therefore you can write a number as `9007199254740993e0` and it will generate the _number_ `9007199254740993e`. We have to interpret that as `9007199254740992`, but `(long) Double.parseDouble("9007199254740993e0")` returns `9007199254740992`.

Using `BigInteger`/`BigDecimal` should be correct, but IMHO pretty expensive. I think it shouldn't be that hard to write our own parser. Something like:

```java
int e = indexOfExponent(str);
if (e < 0) {
  return Long.parse(str);
}
long base = Long.parse(str, 0, e, 10));
int exp = Integer.parseInt(str, e+1, 10));
long powerOfTwo = powerOfTwo(exp); // as suggested in https://stackoverflow.com/questions/46983772/fastest-way-to-obtain-a-power-of-10
long result = verifyOverflow(base * powerOfTen(exp)); // Not sure if checking if the value changed the sign is good enough
```

Alternatively, if we are not sure if that algorithm is correct, we can do:
```java
int e = indexOfExponent(str);
if (e < 0) {
  return Long.parse(str);
}
if (str.contains('.')) {
  throw whatever;
}
return toLong(new BigDecimal(str)); // verifying we fail if the value is not representable as double.
```

Reply:
You're right, I was looking at productions only because the main images don't display properly for me. 
What about the current behavior of parsing numbers with fractional parts ? 
Both 123.456 and 122.11e1 are accepted.

Reply:
You are right, my algorithms are not correct because in order to keep backward compatibility we should round them in the same way double to long is doing, but without losing precision. 

So `123.456` should be transformed into `123`. I think it is acceptable if we fail in case we find a number that is too large or small to be represented as a long

Reply:
I added custom methods for long and 'jsonLong' parsing.